File: leetcode-submissions/2577_Minimum_Time_to_Visit_a_Cell_In_a_Grid.cpp
class Solution {
public:
    #define p pair<int,pair<int,int>>
    int minimumTime(vector<vector<int>>& grid) {
        if(grid[0][1]>1  and grid[1][0]>1){
            return -1;
        }
        int m=grid.size();
        int n=grid[0].size();

        vector<vector<int>>result(m,vector<int>(n,INT_MAX));
        result[0][0]=0;
        priority_queue<p,vector<p>,greater<p>>pq;
        pq.push({0,{0,0}});
        vector<vector<bool>>visited(m,vector<bool>(n,false));
        vector<vector<int>>directions ={{0,1},{1,0},{0,-1},{-1,0}};
        while(not pq.empty()){
         auto current =pq.top();
         pq.pop();
         int time =current.first;
         int i=current.second.first;
         int j=current.second.second ;
         if(visited[i][j]){
            continue;
         }
         if(i==m-1 and j==n-1){
            return result[m-1][n-1];
         }
         visited[i][j]=true;
         for(auto &dir:directions){
            int newRow=i+dir[0];
            int newCol=j+dir[1];
            if(newRow>=0 and newCol>=0 and newCol<n and newRow<m){
                if(grid[newRow][newCol]<=time){
                    pq.push({time+1,{newRow,newCol}});
                    result[newRow][newCol]=time+1;
                }else if((grid[newRow][newCol]-time)%2==0){
                    pq.push({grid[newRow][newCol]+1,{newRow,newCol}});
                    result[newRow][newCol]=grid[newRow][newCol]+1;
                }else{
                    pq.push({grid[newRow][newCol],{newRow,newCol}});
                    result[newRow][newCol]=grid[newRow][newCol];
                }
            }
         }
        }

        return result[m-1][n-1];


    }
};

File: leetcode-submissions/2577_Minimum_Time_to_Visit_a_Cell_In_a_Grid.ts
function minimumTime(grid: number[][]): number {
    if(grid[0][1]>1 && grid[1][0]>1){
        return -1;
    }
    const m:number =grid.length;
    const n:number =grid[0].length;
    const result:number[][]=Array.from({length:m},()=>Array.from({length:n},()=>Infinity));
    const pq:[number ,number,number][]=[];
    const push=(time:number ,row:number ,col:number):void=>{
        pq.push([time,row,col]);
        pq.sort((a,b)=>a[0]-b[0]);

    }
    result[0][0]=0;
    const directions :number[][]=[[0,1],[0,-1],[1,0],[-1,0]];
    const visited:boolean[][]=Array.from({length:m},()=>Array.from({length:n},()=>false));
    push(0,0,0);
    while(pq.length>0){
    const [time ,row,col]=pq.shift();
    pq.unshift();
    if(visited[row][col]){
        continue ;

    }
    if(row==m-1 && col==n-1){
        return result[m-1][n-1];

    }
    for(const dir of directions ){
        let i=row+dir[0];
        let j=col+dir[1];
        if(i<0 || j<0 || j>=n || i>=m){
            continue;

        }
        if(grid[i][j]<=time){
            push(time+1,i,j);
            result[i][j]=time+1;
        }else if ((grid[i][j]-time)%2===0){
            push(grid[i][j]+1,i,j);
            result[i][j]=grid[i][j]+1;
        }else {
            push(grid[i][j],i,j);
            result[i][j]=grid[i][j];

        }

    }


    }
    return result[m-1][n-1] ;


};